File: src/graph/topological-sort/km117.py
from typing import Dict, List


vertexNum, edgeNum = map(int, input().strip().split(" "))

# 入度
inDegree: List[int] = [0 for _ in range(vertexNum)]

edges: Dict[int, List[int]] = dict()

idxQueue: List[int] = []
ans: List[int] = []

for _ in range(edgeNum):
    start, end = map(int, input().strip().split(" "))
    inDegree[end] += 1
    if start in edges:
        edges[start].append(end)
    else:
        edges[start] = [end]

for i in range(vertexNum):
    if inDegree[i] == 0:
        idxQueue.append(i)

while len(idxQueue) > 0:
    head = idxQueue.pop(0)
    ans.append(head)

    dst = edges.get(head)
    if dst is not None:
        for item in dst:
            inDegree[item] -= 1
            if inDegree[item] == 0:
                idxQueue.append(item)

if len(ans) == vertexNum:
    print(" ".join(map(str, ans)))
else:
    print(-1)

File: src/graph/topological-sort/km117.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;

let vertexNum = 0;
let edgeNum = 0;

let inDegree: number[] = [];
const edges: Map<number /** src */, number[] /** dst */> = new Map<
  number, // src
  number[] // dst
>();

const idxQueue: number[] = [];
const ans: number[] = [];

rl.on("line", (line) => {
  lineno++;

  if (lineno === 1) {
    [vertexNum, edgeNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));

    // initialize
    inDegree = new Array(vertexNum).fill(0);
    return;
  }

  const [start, end] = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));
  inDegree[end]++;

  if (edges.has(start)) {
    edges.get(start)?.push(end);
  } else {
    edges.set(start, [end]);
  }

  if (lineno === 1 + edgeNum) {
    rico();
    rl.close();
  }
});

function rico() {
  for (let i = 0; i < vertexNum; i++) {
    if (inDegree[i] === 0) {
      idxQueue.push(i);
    }
  }

  while (idxQueue.length) {
    const head = idxQueue[0];
    idxQueue.shift();
    ans.push(head);

    const dst = edges.get(head);
    if (dst) {
      for (let i = 0; i < dst.length; i++) {
        inDegree[dst[i]]--;
        if (inDegree[dst[i]] === 0) {
          idxQueue.push(dst[i]);
        }
      }
    }
  }

  if (ans.length === vertexNum) {
    console.log(ans.map((item) => item.toString()).join(" "));
  } else {
    console.log(-1);
  }
}
